javatests/arcs/core/analysis/ExpressionClaimDeducerTest.kt

```diff
@@ -0,0 +1,341 @@
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.analysis
+
+import arcs.core.data.expression.PaxelParser
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+@RunWith(JUnit4::class)
+class ExpressionClaimDeducerTest {
+
+  @Test
+  fun literal_number() {
+    val expr = PaxelParser.parse("5")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(Deduction.Empty)
+  }
+
+  @Test
+  fun literal_text() {
+    val expr = PaxelParser.parse("'hello'")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(Deduction.Empty)
+  }
+
+  @Test
+  fun literal_boolean() {
+    val expr = PaxelParser.parse("false")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(Deduction.Empty)
+  }
+
+  @Test
+  fun literal_null() {
+    val expr = PaxelParser.parse("null")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(Deduction.Empty)
+  }
+
+  @Test
+  fun single_variable() {
+    val expr = PaxelParser.parse("x")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Equal("x")
+    )
+  }
+
+  @Test
+  fun sum_variable_literal() {
+    val expr = PaxelParser.parse("x + 1")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Derive(
+        Deduction.Equal("x")
+      )
+    )
+  }
+
+  @Test
+  fun variable_field_access() {
+    val expr = PaxelParser.parse("x.foo")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Equal("x", "foo")
+    )
+  }
+
+  @Test
+  fun variable_field_access_nested() {
+    val expr = PaxelParser.parse("x.foo.bar")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Equal("x", "foo", "bar")
+    )
+  }
+
+  @Test
+  fun variable_field_access_binop() {
+    val expr = PaxelParser.parse("x.foo.bar + y.foo.bar.baz + z.baz.bar")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Derive(
+        listOf("x", "foo", "bar"),
+        listOf("y", "foo", "bar", "baz"),
+        listOf("z", "baz", "bar")
+      )
+    )
+  }
+
+  @Test
+  fun literal_binop() {
+    val expr = PaxelParser.parse("1 + 2 + 3 + 4 + 5")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Empty
+    )
+  }
+
+  @Test
+  fun variable_field_access_binop_literals() {
+    val expr = PaxelParser.parse("x.foo.bar + y.foo.bar.baz + z.baz.bar + 10 + 20")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Derive(
```

Comment:
Something weird is happening here. here `Derive` is composed of `Equals`, but in `variable_field_access_binop` test almost the same expression produced a `Derive` that is composed of paths, not `Equals`.

I don't think `Derive` should be composed of other deductions, only paths should be allowed. There's not need to derive that is composed of other derives as far as I can tell.